libkbfs/block_tree_test.go

```diff
@@ -0,0 +1,65 @@
+package libkbfs
+
+import (
+	"github.com/golang/mock/gomock"
+	"reflect"
+	"testing"
+)
+
+func makePaths(childIndices [][]int,
+	mockBlock BlockWithPtrs) [][]parentBlockAndChildIndex {
+	result := [][]parentBlockAndChildIndex(nil)
+	for _, indexList := range childIndices {
+		path := []parentBlockAndChildIndex(nil)
+		for _, childIndex := range indexList {
+			path = append(path, parentBlockAndChildIndex{
+				pblock:     mockBlock,
+				childIndex: childIndex,
+			})
+		}
+		result = append(result, path)
+	}
+	return result
+}
+
+func TestCheckForHolesAndTruncate(t *testing.T) {
+	// Make a mock block that believes itself to have three children.
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockBlock := NewMockBlockWithPtrs(mockCtrl)
+	mockBlock.EXPECT().NumIndirectPtrs().Return(4).MinTimes(1)
+
+	goodExamples := [][][]int{
+		{{3, 2}, {3, 3}},
+		{{2, 3}, {3, 0}},
+		{{1, 0}, {1, 1}, {1, 2}, {1, 3}, {2, 0}, {2, 1}},
+		{{1, 2, 3}, {1, 3, 0}, {1, 3, 1}, {1, 3, 2}, {1, 3, 3}, {2, 0, 0}},
+	}
+	badExamples := [][][]int{
+		{{3, 3}, {3, 2}},
+		{{1, 3, 0}, {1, 3, 1}, {1, 3, 2}, {2, 0, 3}},
+		{{1, 0}, {1, 1}, {1, 2}, {2, 0}, {2, 1}},
+	}
+	// correctLengths holds the length that each bad example should be truncated
+	// to.
+	correctLengths := []int{1, 3, 3}
+
+	for _, goodEx := range goodExamples {
+		paths := makePaths(goodEx, mockBlock)
+		newPaths := checkForHolesAndTruncate(paths)
+		if !reflect.DeepEqual(paths, newPaths) {
+			t.Errorf("Paths incorrectly truncated from %v to %v",
```

Comment:
Please use `require.True(t, reflect.DeepEqual(paths, newPaths))` instead of the `if` and the `t.Errorf`. (We do this for new tests, but haven't fixed all the old tests yet.)  There's an optional third parameter for messages as well, if you really want to print `paths` and `newPaths` when the check fails.

Reply:
I do really want to print them, it was useful when debugging the originally incorrect implementation.